server-spi-private/src/main/java/org/keycloak/policy/HistoryPasswordPolicyProvider.java

```diff
@@ -53,21 +54,23 @@ public PolicyError validate(RealmModel realm, UserModel user, String password) {
         PasswordPolicy policy = session.getContext().getRealm().getPasswordPolicy();
         int passwordHistoryPolicyValue = policy.getPolicyConfig(PasswordPolicy.PASSWORD_HISTORY_ID);
         if (passwordHistoryPolicyValue != -1) {
-            List<CredentialModel> storedPasswords = session.userCredentialManager().getStoredCredentialsByType(realm, user, PasswordCredentialModel.TYPE);
-            for (CredentialModel cred : storedPasswords) {
-                PasswordCredentialModel passwordCredential = PasswordCredentialModel.createFromCredentialModel(cred);
+            List<PasswordCredentialModel> passwordCredentials = session.userCredentialManager()
+                    .getStoredCredentialsByTypeStream(realm, user, PasswordCredentialModel.TYPE)
+                    .map(PasswordCredentialModel::createFromCredentialModel)
+                    .collect(Collectors.toList());
+            for (PasswordCredentialModel passwordCredential : passwordCredentials) {
                 PasswordHashProvider hash = session.getProvider(PasswordHashProvider.class, passwordCredential.getPasswordCredentialData().getAlgorithm());
-                if (hash == null) continue;
-                if (hash.verify(password, passwordCredential)) {
+                if (hash != null && hash.verify(password, passwordCredential)) {
                     return new PolicyError(ERROR_MESSAGE, passwordHistoryPolicyValue);
                 }
             }
 
             if (passwordHistoryPolicyValue > 0) {
-                List<CredentialModel> passwordHistory = session.userCredentialManager().getStoredCredentialsByType(realm, user, PasswordCredentialModel.PASSWORD_HISTORY);
-                List<CredentialModel> recentPasswordHistory = getRecent(passwordHistory, passwordHistoryPolicyValue - 1);
-                for (CredentialModel cred : recentPasswordHistory) {
-                    PasswordCredentialModel passwordCredential = PasswordCredentialModel.createFromCredentialModel(cred);
+                Stream<CredentialModel> passwordHistory = session.userCredentialManager().getStoredCredentialsByTypeStream(realm, user, PasswordCredentialModel.PASSWORD_HISTORY);
+                List<PasswordCredentialModel> recentPasswordHistory = this.getRecent(passwordHistory, passwordHistoryPolicyValue - 1)
+                        .map(PasswordCredentialModel::createFromCredentialModel)
+                        .collect(Collectors.toList());
+                for (PasswordCredentialModel passwordCredential : recentPasswordHistory) {
```

Comment:
Maybe something similar could be used as in suggestion above.

Reply:
Definitely!